mirgecom/artificial_viscosity.py

```diff
@@ -152,6 +159,25 @@ def _facial_flux_r(discr, r_tpair):
     return discr.project(r_tpair.dd, "all_faces", flux_out)
 
 
+class AVBoundaryInterface(metaclass=ABCMeta):
+    r"""Abstract interface to boundary treatments used by artificial viscosity.
+
+    This is the interface that must be implemented by any boundary treatment
+    device used by the artificial viscosity operator (:func:`av_operator`).
+
+    .. automethod:: get_scalar_flux
+    .. automethod:: get_diffusion_flux
+    """
+
+    @abstractmethod
+    def get_scalar_flux(self, discr, btag, u, **kwargs):
+        """Get the scalar flux of *u* components through boundary faces in *btag*."""
```

Comment:
> It is the flux of each scalar component of _q_ at the boundary face

In reference to what equation? `get_scalar_flux` doesn't seem very descriptive, and it is potentially confusing in a fluid context (as you just observed on Slack).

Reply:
`get_scalar_flux` is independent of equation.  It is simply the returns the flux of the scalar (or scalar components of vector) *u*.   In the context of the BR1 approach to NS, it would be equation BR1.13 the scalar boundary flux part of BR1.12.

Reply:
> `get_scalar_flux` is independent of equation. It is simply the returns the flux of the scalar (or scalar components of vector) _u_.

? I'm still not really grasping what you're saying here.

In order to talk about the flux of u, don't you need to have a flux function that comes from the equation/operator? Or are you just saying that the form of the flux function is `u`? (As far as I'm aware these are not interchangeable.) If it's the latter case, I think I would object to calling it `get_scalar_flux`.

Reply:
> > `get_scalar_flux` is independent of equation. It is simply the returns the flux of the scalar (or scalar components of vector) _u_.
> 
> ? I'm still not really grasping what you're saying here.
> 
> In order to talk about the flux of u, don't you need to have a flux function that comes from the equation/operator? Or are you just saying that the form of the flux function is `u`? (As far as I'm aware these are not interchangeable.) If it's the latter case, I think I would object to calling it `get_scalar_flux`.

The flux of a thing is a thing in-and-of-itself, and is independent of how it might be used. This flux would be used any time I want to integrate *u* over the surface area, not just for a gradient, and not just for any particular governing equation.

Reply:
I'll admit I'm as confused as @majosm. To my mind, the flux function comes from/is part of the PDE. The solution vector itself (or its components) doesn't have a flux (or, it has many... the viscous and inviscid ones, their sum...).

Reply:
In my view, the inviscid and viscous fluxes are transport fluxes where material is carried along by the respective transport phenomena. That is, they are inherent to the system we are modeling, *not* our discretization of it.   

Enter discretization; where we need to conduct surface integrals of scalar and vector quantities over the surface of the element.  For transport phenomena, an arbitrary scalar material "u" transported in some direction (vector) "v", the flux across the face would be (uv.dot.nhat) in the nhat direction.  (nhat being the face normal).   

From my perspective, the content of "u", direction and magnitude of "v" are due to the transport phenomenon, and calculating the resulting flux across the discrete face is the same operation regardless of what is driving transport.   

For us, the surface integral of this quantity:

scalar(U) * nhat 

appears several times in different equations. This is what I am calling the scalar flux function, regardless of the actual content of U. 

We integrate these vector quantities over the element boundary too:

(F .dot. nhat)

Here, F can be any vector field that we want to integrate over the element boundary. If the content of F is the inviscid flux, then we get the flux due to advection, if the viscous flux then diffusive and so forth.  But regardless of content we perform the operation the same.

(when I say *the same* I mean up to numerical flux where we replace (Unhat or F.dot.nhat) with a numerical flux because two elements do not agree on U or F at the face. The user is (should be) free to choose the numerical flux function used for each integral in our code.)

Reply:
If we want to rename this function to something else that is less confusing - then let's just do that.  If we want to keep talking about why I named it the way I did (and/or why that is confusing, or wrong) then we can do that any time, and I'm happy to and very interested. But, let's just land on something that works for everyone for right now and move on.

Reply:
I think the trouble with trying to generalize something like that is that the boundary condition won't necessarily know which scalar is being passed in (and if it's enforcing a BC, it needs to know that).

I had an idea last night that I want to try out. I'll get back to you in a bit.

Reply:
> I think the trouble with trying to generalize something like that is that the boundary condition won't necessarily know which scalar is being passed in (and if it's enforcing a BC, it needs to know that).

It really isn't generalized as-is. The BCs still need to be applied, and the BC on the inside already knows how to pick *u* apart into whatever pieces are needed to apply boundary conditions. Fluid BC's, for example, use `split_conserved`. Once the U_bc are known, the generic U*normal (what I called scalar flux - simply for lack of a better name) is returned. 

The fluid-equivalent function I called `get_q_flux` referring to Q as the vector of conserved fluid quantities. The BCs inherit both interfaces, and set the AV-specific `get_scalar_flux` to `get_q_flux` so that the AV operator gets BCs consistent with the fluid boundary treatment.

Reply:
You're not going to like this, but: I'm going to suggest `get_gradient_flux` again.

Here's why: [link](https://github.com/majosm/mirgecom/commit/1d2e04b2380aa522d917c607b5fd0cc4940747b8)

In this code I added a `gradient_operator` and reimplemented diffusion using it. You can see that here `get_gradient_flux` provides a nice symmetry: all of the flux retrieval routines take the form `get_<operator-acting-on-state>_flux`. This gradient operator is something I could see us adding in the future (maybe along with a `divergence_operator`/`DivergenceBoundaryInterface`; then we might be able to get rid of `DiffusionBoundaryInterface`/`AVBoundaryInterface`?). I definitely wouldn't recommend piling this on top of NS right now, but I figure there's no harm in getting the flux name right in the meantime.

As for ambiguity with the temperature gradient, I think `get_temperature_gradient_flux` would fit into this scheme, in the sense that you can think of the temperature gradient as an operator `TempGrad(q)`.

Reply:
> You're not going to like this, but: I'm going to suggest `get_gradient_flux` again.

I don't dislike this at all. In the context of a gradient-specific operator, I think the `get_gradient_flux` naming is descriptive enough, and just great.

Re: operator-specific flux retrieval, you are probably aware that NS uses:
- `get_inviscid_flux`
- `get_viscous_flux`
- `get_temperature_flux` -> `TempGrad(q)`
- `get_q_flux` -> `gradient_operator(q)`

I only point this out as there is no `get_ns_flux`, for example.  The inviscid and viscous parts are treated separate. I don't think that is an issue here in what you've laid out.

Just so I am clear, what I think you are suggesting is that we go ahead and adopt this language for our flux routines looking forward to the addition of the gradient, and potentially other differential operators in the near future.

Reply:
> Re: operator-specific flux retrieval, you are probably aware that NS uses:
> 
> * `get_inviscid_flux`
> * `get_viscous_flux`
> * `get_temperature_flux` -> `TempGrad(q)`
> * `get_q_flux` -> `gradient_operator(q)`
> 
> I only point this out as there is no `get_ns_flux`, for example. The inviscid and viscous parts are treated separate. I don't think that is an issue here in what you've laid out.

Right, I think this is fine. You can still think of there being inviscid/viscous operators even if they're not explicitly laid out that way in the code.

I'm also leaning a little bit towards saying we can just drop the `EulerBoundaryInterface`/`NSBoundaryInterface` and use `InviscidBoundaryInterface`/`ViscousBoundaryInterface` directly. Probably the same for `AVBoundaryInterface` once diffusion-based AV is added.

> Just so I am clear, what I think you are suggesting is that we go ahead and adopt this language for our flux routines looking forward to the addition of the gradient, and potentially other differential operators in the near future.

:thumbsup:

Reply:
Thanks, Matt. This all sounds great.  I'll massage it into place as you describe, and we can go around again :).